Function luaL_openlib was replaced by luaL_register in Lua 5.1

##### Minetest version

```
5.1.4
```

##### OS / Hardware

Operating system: Gentoo Linux

##### Summary

"Function `luaL_openlib` was replaced by `luaL_register`" in Lua 5.1: https://www.lua.org/manual/5.1/manual.html#7.3
It is only available in the backwards compatibility / deprecation layer in Lua 5.1.

Note that "Function `luaL_register` is deprecated. Use `luaL_setfuncs` so that your module does not create globals. (Modules are not expected to set global variables anymore.)" in Lua 5.2: https://www.lua.org/manual/5.2/manual.html#8.3

An example for how to replace the calls to `luaL_openlib` is given e.g. on StackOverflow: https://stackoverflow.com/questions/19041215/lual-openlib-replacement-for-lua-5-2

The biggest change will probably be in the consumers of your modules, which now need to store the method tables themselves instead of relying on globals being set for them.

##### Steps to reproduce

1. Build Lua 5.1 without deprecated features.
2. Try to build Minetest 5.4.1.
3. Observe `.../minetest-5.4.1/src/script/lua_api/l_nodetimer.cpp:125:9: error: ‘luaL_openlib’ was not declared in this scope`.

For detailed logs please see https://bugs.gentoo.org/791928

--- a/builtin/game/forceloading.lua
+++ b/builtin/game/forceloading.lua
@@ -86,12 +86,6 @@ local function read_file(filename)
 	return core.deserialize(t) or {}
 end
 
-local function write_file(filename, table)
-	local f = io.open(filename, "w")
-	f:write(core.serialize(table))
-	f:close()
-end
-
 blocks_forceloaded = read_file(wpath.."/force_loaded.txt")
 for _, __ in pairs(blocks_forceloaded) do
 	total_forceloaded = total_forceloaded + 1
@@ -106,7 +100,8 @@ end)
 
 -- persists the currently forceloaded blocks to disk
 local function persist_forceloaded_blocks()
-	write_file(wpath.."/force_loaded.txt", blocks_forceloaded)
+	local data = core.serialize(blocks_forceloaded)
+	core.safe_file_write(wpath.."/force_loaded.txt", data)
 end
 
 -- periodical forceload persistence

--- a/cmake/Modules/FindLuaJIT.cmake
+++ b/cmake/Modules/FindLuaJIT.cmake
@@ -1,8 +1,8 @@
 # Locate LuaJIT library
 # This module defines
 #  LUAJIT_FOUND, if false, do not try to link to Lua
+#  LUA_LIBRARY, where to find the lua library
 #  LUA_INCLUDE_DIR, where to find lua.h
-#  LUA_VERSION_STRING, the version of Lua found (since CMake 2.8.8)
 #
 # This module is similar to FindLua51.cmake except that it finds LuaJit instead.
 
@@ -44,19 +44,10 @@ else()
 	)
 endif()
 
-
-IF(LUA_INCLUDE_DIR AND EXISTS "${LUA_INCLUDE_DIR}/luajit.h")
-	FILE(STRINGS "${LUA_INCLUDE_DIR}/luajit.h" lua_version_str REGEX "^#define[ \t]+LUA_RELEASE[ \t]+\"LuaJIT .+\"")
-
-	STRING(REGEX REPLACE "^#define[ \t]+LUA_RELEASE[ \t]+\"LuaJIT ([^\"]+)\".*" "\\1" LUA_VERSION_STRING "${lua_version_str}")
-	UNSET(lua_version_str)
-ENDIF()
-
 INCLUDE(FindPackageHandleStandardArgs)
 # handle the QUIETLY and REQUIRED arguments and set LUAJIT_FOUND to TRUE if
-# all listed variables are TRUE
+# all listed variables exist
 FIND_PACKAGE_HANDLE_STANDARD_ARGS(LuaJIT
-	REQUIRED_VARS LUA_LIBRARY LUA_INCLUDE_DIR
-	VERSION_VAR LUA_VERSION_STRING)
+	REQUIRED_VARS LUA_LIBRARY LUA_INCLUDE_DIR)
 
-MARK_AS_ADVANCED(LUA_INCLUDE_DIR LUA_LIBRARY LUA_MATH_LIBRARY)
+MARK_AS_ADVANCED(LUA_INCLUDE_DIR LUA_LIBRARY)

--- a/src/client/content_cao.cpp
+++ b/src/client/content_cao.cpp
@@ -346,18 +346,6 @@ void GenericCAO::initialize(const std::string &data)
 	infostream<<"GenericCAO: Got init data"<<std::endl;
 	processInitData(data);
 
-	if (m_is_player) {
-		// Check if it's the current player
-		LocalPlayer *player = m_env->getLocalPlayer();
-		if (player && strcmp(player->getName(), m_name.c_str()) == 0) {
-			m_is_local_player = true;
-			m_is_visible = false;
-			player->setCAO(this);
-
-			m_prop.show_on_minimap = false;
-		}
-	}
-
 	m_enable_shaders = g_settings->getBool("enable_shaders");
 }
 
@@ -379,6 +367,16 @@ void GenericCAO::processInitData(const std::string &data)
 	m_position = readV3F32(is);
 	m_rotation = readV3F32(is);
 	m_hp = readU16(is);
+
+	if (m_is_player) {
+		// Check if it's the current player
+		LocalPlayer *player = m_env->getLocalPlayer();
+		if (player && strcmp(player->getName(), m_name.c_str()) == 0) {
+			m_is_local_player = true;
+			m_is_visible = false;
+			player->setCAO(this);
+		}
+	}
 
 	const u8 num_messages = readU8(is);
 

--- a/src/database/database-postgresql.cpp
+++ b/src/database/database-postgresql.cpp
@@ -39,20 +39,24 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "remoteplayer.h"
 #include "server/player_sao.h"
 
-Database_PostgreSQL::Database_PostgreSQL(const std::string &connect_string) :
+Database_PostgreSQL::Database_PostgreSQL(const std::string &connect_string,
+	const char *type) :
 	m_connect_string(connect_string)
 {
 	if (m_connect_string.empty()) {
-		throw SettingNotFoundException(
-			"Set pgsql_connection string in world.mt to "
+		// Use given type to reference the exact setting in the error message
+		std::string s = type;
+		std::string msg =
+			"Set pgsql" + s + "_connection string in world.mt to "
 			"use the postgresql backend\n"
 			"Notes:\n"
-			"pgsql_connection has the following form: \n"
-			"\tpgsql_connection = host=127.0.0.1 port=5432 user=mt_user "
-			"password=mt_password dbname=minetest_world\n"
+			"pgsql" + s + "_connection has the following form: \n"
+			"\tpgsql" + s + "_connection = host=127.0.0.1 port=5432 "
+			"user=mt_user password=mt_password dbname=minetest" + s + "\n"
 			"mt_user should have CREATE TABLE, INSERT, SELECT, UPDATE and "
-			"DELETE rights on the database.\n"
-			"Don't create mt_user as a SUPERUSER!");
+			"DELETE rights on the database. "
+			"Don't create mt_user as a SUPERUSER!";
+		throw SettingNotFoundException(msg);
 	}
 }
 
@@ -166,7 +170,7 @@ void Database_PostgreSQL::rollback()
 }
 
 MapDatabasePostgreSQL::MapDatabasePostgreSQL(const std::string &connect_string):
-	Database_PostgreSQL(connect_string),
+	Database_PostgreSQL(connect_string, ""),
 	MapDatabase()
 {
 	connectToDatabase();
@@ -315,7 +319,7 @@ void MapDatabasePostgreSQL::listAllLoadableBlocks(std::vector<v3s16> &dst)
  * Player Database
  */
 PlayerDatabasePostgreSQL::PlayerDatabasePostgreSQL(const std::string &connect_string):
-	Database_PostgreSQL(connect_string),
+	Database_PostgreSQL(connect_string, "_player"),
 	PlayerDatabase()
 {
 	connectToDatabase();
@@ -637,7 +641,8 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
 }
 
 AuthDatabasePostgreSQL::AuthDatabasePostgreSQL(const std::string &connect_string) :
-		Database_PostgreSQL(connect_string), AuthDatabase()
+	Database_PostgreSQL(connect_string, "_auth"),
+	AuthDatabase()
 {
 	connectToDatabase();
 }

--- a/src/database/database-postgresql.h
+++ b/src/database/database-postgresql.h
@@ -29,7 +29,7 @@ class Settings;
 class Database_PostgreSQL: public Database
 {
 public:
-	Database_PostgreSQL(const std::string &connect_string);
+	Database_PostgreSQL(const std::string &connect_string, const char *type);
 	~Database_PostgreSQL();
 
 	void beginSave();

--- a/src/gui/guiFormSpecMenu.cpp
+++ b/src/gui/guiFormSpecMenu.cpp
@@ -1577,11 +1577,10 @@ void GUIFormSpecMenu::createTextField(parserData *data, FieldSpec &spec,
 		}
 
 		e->setNotClipped(style.getBool(StyleSpec::NOCLIP, false));
-		e->setDrawBorder(style.getBool(StyleSpec::BORDER, true));
 		e->setOverrideColor(style.getColor(StyleSpec::TEXTCOLOR, video::SColor(0xFFFFFFFF)));
-		if (style.get(StyleSpec::BGCOLOR, "") == "transparent") {
-			e->setDrawBackground(false);
-		}
+		bool border = style.getBool(StyleSpec::BORDER, true);
+		e->setDrawBorder(border);
+		e->setDrawBackground(border);
 		e->setOverrideFont(style.getFont());
 
 		e->drop();

--- a/src/script/lua_api/l_areastore.cpp
+++ b/src/script/lua_api/l_areastore.cpp
@@ -372,7 +372,7 @@ void LuaAreaStore::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Can be created from Lua (AreaStore())

--- a/src/script/lua_api/l_camera.cpp
+++ b/src/script/lua_api/l_camera.cpp
@@ -219,7 +219,7 @@ void LuaCamera::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 }
 

--- a/src/script/lua_api/l_env.cpp
+++ b/src/script/lua_api/l_env.cpp
@@ -228,7 +228,7 @@ void LuaRaycast::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);

--- a/src/script/lua_api/l_inventory.cpp
+++ b/src/script/lua_api/l_inventory.cpp
@@ -456,7 +456,7 @@ void InvRef::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Cannot be created from Lua

--- a/src/script/lua_api/l_item.cpp
+++ b/src/script/lua_api/l_item.cpp
@@ -483,7 +483,7 @@ void LuaItemStack::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Can be created from Lua (ItemStack(itemstack or itemstring or table or nil))

--- a/src/script/lua_api/l_itemstackmeta.cpp
+++ b/src/script/lua_api/l_itemstackmeta.cpp
@@ -114,7 +114,7 @@ void ItemStackMetaRef::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Cannot be created from Lua

--- a/src/script/lua_api/l_localplayer.cpp
+++ b/src/script/lua_api/l_localplayer.cpp
@@ -446,7 +446,7 @@ void LuaLocalPlayer::Register(lua_State *L)
 
 	lua_pop(L, 1); // Drop metatable
 
-	luaL_openlib(L, 0, methods, 0); // fill methodtable
+	luaL_register(L, nullptr, methods); // fill methodtable
 	lua_pop(L, 1);			// Drop methodtable
 }
 

--- a/src/script/lua_api/l_minimap.cpp
+++ b/src/script/lua_api/l_minimap.cpp
@@ -211,7 +211,7 @@ void LuaMinimap::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 }
 

--- a/src/script/lua_api/l_modchannels.cpp
+++ b/src/script/lua_api/l_modchannels.cpp
@@ -107,7 +107,7 @@ void ModChannelRef::Register(lua_State *L)
 
 	lua_pop(L, 1); // Drop metatable
 
-	luaL_openlib(L, 0, methods, 0); // fill methodtable
+	luaL_register(L, nullptr, methods); // fill methodtable
 	lua_pop(L, 1);			// Drop methodtable
 }
 

--- a/src/script/lua_api/l_nodemeta.cpp
+++ b/src/script/lua_api/l_nodemeta.cpp
@@ -234,7 +234,7 @@ void NodeMetaRef::RegisterCommon(lua_State *L)
 void NodeMetaRef::Register(lua_State *L)
 {
 	RegisterCommon(L);
-	luaL_openlib(L, 0, methodsServer, 0);  // fill methodtable
+	luaL_register(L, nullptr, methodsServer);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 }
 
@@ -260,7 +260,7 @@ const luaL_Reg NodeMetaRef::methodsServer[] = {
 void NodeMetaRef::RegisterClient(lua_State *L)
 {
 	RegisterCommon(L);
-	luaL_openlib(L, 0, methodsClient, 0);  // fill methodtable
+	luaL_register(L, nullptr, methodsClient);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 }
 

--- a/src/script/lua_api/l_nodetimer.cpp
+++ b/src/script/lua_api/l_nodetimer.cpp
@@ -122,7 +122,7 @@ void NodeTimerRef::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Cannot be created from Lua

--- a/src/script/lua_api/l_noise.cpp
+++ b/src/script/lua_api/l_noise.cpp
@@ -122,7 +122,7 @@ void LuaPerlinNoise::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);
@@ -380,7 +380,7 @@ void LuaPerlinNoiseMap::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);
@@ -485,7 +485,7 @@ void LuaPseudoRandom::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);
@@ -584,7 +584,7 @@ void LuaPcgRandom::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);
@@ -699,7 +699,7 @@ void LuaSecureRandom::Register(lua_State *L)
 
 	lua_pop(L, 1);
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	lua_register(L, className, create_object);

--- a/src/script/lua_api/l_object.cpp
+++ b/src/script/lua_api/l_object.cpp
@@ -2311,7 +2311,7 @@ void ObjectRef::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 }
 

--- a/src/script/lua_api/l_playermeta.cpp
+++ b/src/script/lua_api/l_playermeta.cpp
@@ -97,7 +97,7 @@ void PlayerMetaRef::Register(lua_State *L)
 
 	lua_pop(L, 1); // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);
+	luaL_register(L, nullptr, methods);
 	lua_pop(L, 1);
 
 	// Cannot be created from Lua

--- a/src/script/lua_api/l_settings.cpp
+++ b/src/script/lua_api/l_settings.cpp
@@ -334,7 +334,7 @@ void LuaSettings::Register(lua_State* L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Can be created from Lua (Settings(filename))

--- a/src/script/lua_api/l_storage.cpp
+++ b/src/script/lua_api/l_storage.cpp
@@ -110,7 +110,7 @@ void StorageRef::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 }
 

--- a/src/script/lua_api/l_vmanip.cpp
+++ b/src/script/lua_api/l_vmanip.cpp
@@ -450,7 +450,7 @@ void LuaVoxelManip::Register(lua_State *L)
 
 	lua_pop(L, 1);  // drop metatable
 
-	luaL_openlib(L, 0, methods, 0);  // fill methodtable
+	luaL_register(L, nullptr, methods);  // fill methodtable
 	lua_pop(L, 1);  // drop methodtable
 
 	// Can be created from Lua (VoxelManip())

